lib/flowViewport/API/perViewportSceneIndicesData/fvpFilteringSceneIndicesChainManager.cpp

```diff
@@ -0,0 +1,170 @@
+    
+    if (viewportInformationAndSceneIndicesPerViewportData.GetLastFilteringSceneIndexOfTheChain() == nullptr){
+        TF_CODING_ERROR("viewportInformationAndSceneIndicesPerViewportData->GetLastFilteringSceneIndexOfTheChain() == nullptr is invalid here");
+        return nullptr;
+    }
+
+    //Add the last element of the filtering scene indices chain to the render index
+    return viewportInformationAndSceneIndicesPerViewportData.GetLastFilteringSceneIndexOfTheChain();
+}
+
+void FilteringSceneIndicesChainManager::destroyFilteringSceneIndicesChain(ViewportInformationAndSceneIndicesPerViewportData& viewportInformationAndSceneIndicesPerViewportData)
+{
+    HdSceneIndexBaseRefPtr& lastSceneIndexOfTheChain = viewportInformationAndSceneIndicesPerViewportData.GetLastFilteringSceneIndexOfTheChain();
+    if (nullptr == lastSceneIndexOfTheChain){
+        return;
+    }
+
+    auto renderIndexProxy = viewportInformationAndSceneIndicesPerViewportData.GetRenderIndexProxy();
+    TF_AXIOM(renderIndexProxy);
+    auto renderIndex = renderIndexProxy->GetRenderIndex();
+    TF_AXIOM(renderIndex);
+    renderIndex->RemoveSceneIndex(lastSceneIndexOfTheChain);//Remove the whole chain from the render index
+
+    //Remove a ref on it which should cascade the same on its references
+    lastSceneIndexOfTheChain.Reset();
+}
+
+void FilteringSceneIndicesChainManager::updateFilteringSceneIndicesChain(const std::string& rendererDisplayNames)
+{
+    /*  rendererDisplayName is a string containing either FvpViewportAPITokens->allRenderers meaning this should apply to all renderers 
+    *   or it contains one or more renderers display names such as ("GL, Arnold") and in this case we must update 
+    *   only the viewports filtering scene indices chain which are using this renderer.
+    */
+
+    ViewportInformationAndSceneIndicesPerViewportDataSet& viewportInformationAndSceneIndicesPerViewportDataSet=
+            ViewportInformationAndSceneIndicesPerViewportDataManager::Get().GetAllViewportInfoAndData();
+
+    for (auto& viewportInformationAndSceneIndicesPerViewportData : viewportInformationAndSceneIndicesPerViewportDataSet){
+        
+        //Check the renderer name
+        const std::string& rendererDisplayName = viewportInformationAndSceneIndicesPerViewportData.GetViewportInformation()._rendererName;
+        if ( (FvpViewportAPITokens->allRenderers != rendererDisplayNames) && (! rendererDisplayName.empty()) ){
+            //Filtering per renderer is applied
+            if (std::string::npos == rendererDisplayNames.find(rendererDisplayName)){
+                continue; //Ignore this filtering scene indices chain since the renderer is different
+            }
+        }
+
+        ViewportInformationAndSceneIndicesPerViewportData& nonConstViewportInformationAndSceneIndicesPerViewportData = const_cast<ViewportInformationAndSceneIndicesPerViewportData&>(viewportInformationAndSceneIndicesPerViewportData);
```

Comment:
Because the objects which are in a std::set are supposed to be const. 
I would honestly prefer to use vector since even if search is slower we will have in this set/vector only the size of all viewports used in a DCC which is probably less than 20 anyway. But that was your suggestion in one of the first code review.

Reply:
Agree then, if the elements of the set must have non-const methods called on them, either switch to a map of non-const, or a vector.